### PR TITLE
Client Side API validations (Old)

### PR DESCRIPTION
**Pull request by RidhamShah**
Created clientSide API's body validation for all versions.

### CODE REVIEW

**Comment by danoswaltCL** [comment]
✨ ✨ ✨ ✨  hooray! will we be able to potentially merge in the main within-subjects PR before this one? then this one can get tested and merged against dev separately since it's a bit different focus.

File: backend/packages/Upgrade/src/api/controllers/validators/ExperimentUserValidator.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import { IsNotEmpty, IsOptional, IsString, ValidationOptions, isObject, registerDecorator } from 'class-validator';
+
+export const IsRecord = (validationOptions?: ValidationOptions) => {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/ExperimentUserValidator.ts:3]: IsRecord is too generic name for the specific Group. Better name should be IsGroupRecord

File: backend/packages/Upgrade/src/api/controllers/validators/ExperimentUserValidator.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+            return true;
+          });
+        },
+      },
+    });
+  };
+};
+
+export const IsStringRecord = (validationOptions?: ValidationOptions) => {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/ExperimentUserValidator.ts:34]: IsStringRecord -> IsWorkingGroupRecord

File: backend/packages/Upgrade/src/api/controllers/validators/LogValidator.ts [hunk 3]
```diff
@@ -1,14 +1,57 @@
...
+  groupUniquifier: string;
+
+  @IsObject()
+  attributes: any;
+}
+
+class ILogMetrics {
+  @IsOptional()
+  @IsObject()
+  attributes: any;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/LogValidator.ts:24]: If this is optional can you change it to attributes?:any. And can we replace the any with Record<string, number|string>. I don't think we need any other data type in the value other than number and string.

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.ts [hunk 4]
```diff
@@ -1,22 +1,29 @@
...
   @IsNotEmpty()
   @IsDefined()
+  @IsString()
   public userId: string;
 
-  @IsNotEmpty()
-  @IsDefined()
+  @IsOptional()
+  @IsString()
   public condition: string | null;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.ts:20]: For optional change the type to condition?: string;

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.ts [hunk 5]
```diff
@@ -1,22 +1,29 @@
...
 
+  @IsUUID()
+  @IsOptional()
   public experimentId: string | null;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.ts:24]: Change it to experimentId?: string

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v1.ts [hunk 6]
```diff
@@ -1,22 +1,29 @@
...
   @IsNotEmpty()
   @IsDefined()
+  @IsString()
   public userId: string;
 
-  @IsNotEmpty()
-  @IsDefined()
+  @IsOptional()
+  @IsString()
   public condition: string | null;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v1.ts:20]: Change here

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v1.ts [hunk 7]
```diff
@@ -1,22 +1,29 @@
...
 
+  @IsUUID()
+  @IsOptional()
   public experimentId: string | null;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v1.ts:24]: Change here

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts [hunk 8]
```diff
@@ -1,21 +1,135 @@
...
+class AssignedFactor {
+  @IsDefined()
+  @IsString()
+  level: string;
+
+  @ValidateIf((object, value) => value !== null)
+  @IsObject()
+  @ValidateNested()
+  @Type(() => Payload)
+  payload: Payload | null;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts:70]: Can you use isOptional here?

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts [hunk 9]
```diff
@@ -1,21 +1,135 @@
...
+  @IsDefined()
+  @IsNotEmpty()
+  @IsString()
+  conditionCode: string;
+
+  @IsObject()
+  @ValidateNested()
+  @ValidateIf((object, value) => value !== null)
+  @Type(() => Payload)
+  payload: Payload | null;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts:87]: Can you use isOptional here?

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts [hunk 10]
```diff
@@ -1,21 +1,135 @@
-import { IsNotEmpty, IsDefined } from 'class-validator';
-import { MARKED_DECISION_POINT_STATUS, PAYLOAD_TYPE } from 'upgrade_types';
+import { Type } from 'class-transformer';
+import { IsNotEmpty, IsDefined, IsString, IsOptional, IsEnum, IsObject, ValidateIf, ValidateNested, ValidationOptions, registerDecorator } from 'class-validator';
+import { EXPERIMENT_TYPE, MARKED_DECISION_POINT_STATUS, PAYLOAD_TYPE } from 'upgrade_types';
+
+export const IsAssignedFactorRecord = (validationOptions?: ValidationOptions) => {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts:5]: You can remove export if this decorator is not used outside this file.

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v5.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
+import { Type } from 'class-transformer';
+import { IsNotEmpty, IsDefined, IsString, IsOptional, IsEnum, IsObject, ValidateNested, ValidateIf, ValidationOptions, registerDecorator } from 'class-validator';
+import { EXPERIMENT_TYPE, MARKED_DECISION_POINT_STATUS, PAYLOAD_TYPE } from 'upgrade_types';
+
+export const IsAssignedFactorRecord = (validationOptions?: ValidationOptions) => {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v5.ts:5]: Remove export if unnecesssary.

File: backend/packages/Upgrade/src/api/controllers/validators/MetricValidator.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+      validator: {
+        validate(value: any) {
+          return validateMetricUnit(value)
+        },
+      },
+    });
+  };
+};
+
+function validateMetricUnit(data: any) {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/controllers/validators/MetricValidator.ts:24]: any can be replaced with unknown

File: backend/packages/Upgrade/src/api/models/MonitoredDecisionPointLog.ts [hunk 13]
```diff
@@ -7,6 +7,16 @@ export class MonitoredDecisionPointLog extends BaseModel {
   @PrimaryGeneratedColumn()
   public id: string;
 
+  @Column({
+    nullable: true,
+  })
+  public condition: string;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/models/MonitoredDecisionPointLog.ts:13]: Change to public condition: string | null; as nullable is true

File: backend/packages/Upgrade/src/api/models/MonitoredDecisionPointLog.ts [hunk 14]
```diff
@@ -7,6 +7,16 @@ export class MonitoredDecisionPointLog extends BaseModel {
...
+
+  @Column({
+    nullable: true,
+  })
+  public uniquifier: string;
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/models/MonitoredDecisionPointLog.ts:18]: Change here also

File: backend/packages/Upgrade/src/api/repositories/AnalyticsRepository.ts [hunk 15]
```diff
@@ -114,6 +117,92 @@ export class AnalyticsRepository {
           .execute(),
       ]);
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/repositories/AnalyticsRepository.ts:119]: Why these changes are the part of this PR?

File: backend/packages/Upgrade/src/api/controllers/validators/ExperimentUserValidator.ts (hunk 1 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/ExperimentUserValidator.ts]: Code updated

File: backend/packages/Upgrade/src/api/controllers/validators/ExperimentUserValidator.ts (hunk 2 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/ExperimentUserValidator.ts]: Renamed code

File: backend/packages/Upgrade/src/api/controllers/validators/LogValidator.ts (hunk 3 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/LogValidator.ts]: Code updated

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.ts (hunk 4 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.ts]: Code updated

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.ts (hunk 5 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.ts]: Code updated

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v1.ts (hunk 6 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v1.ts]: Code updated

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v1.ts (hunk 7 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v1.ts]: Code updated

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts (hunk 8 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts:70]: Actually, we don't give payload as an optional value in the code. We pass the payload object or null if not found.

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts (hunk 10 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v4.ts]: Code updated

File: backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v5.ts (hunk 11 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/MarkExperimentValidator.v5.ts]: Code updated

File: backend/packages/Upgrade/src/api/controllers/validators/MetricValidator.ts (hunk 12 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/controllers/validators/MetricValidator.ts]: Code updated

File: backend/packages/Upgrade/src/api/models/MonitoredDecisionPointLog.ts (hunk 13 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/models/MonitoredDecisionPointLog.ts]: Code updated

File: backend/packages/Upgrade/src/api/models/MonitoredDecisionPointLog.ts (hunk 14 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/models/MonitoredDecisionPointLog.ts]: Code updated

**Comment by RidhamShah** [comment]
This PR changes are shifted to another [PR](https://github.com/CarnegieLearningWeb/UpGrade/pull/934) and this PR is not needed.
